atom/browser/ui/win/atom_desktop_native_widget_aura.cc

```diff
@@ -3,12 +3,16 @@
 // found in the LICENSE file.
 
 #include "atom/browser/ui/win/atom_desktop_native_widget_aura.h"
+#include "ui/views/corewm/tooltip_controller.h"
+#include "ui/wm/public/tooltip_client.h"
 
 namespace atom {
 
 AtomDesktopNativeWidgetAura::AtomDesktopNativeWidgetAura(
     views::internal::NativeWidgetDelegate* delegate)
     : views::DesktopNativeWidgetAura(delegate) {
+  // This is to enable the override of OnWindowActivated
+  wm::SetActivationChangeObserver(GetNativeWindow(), this);
```

Comment:
From the libcc API, it looks like we can only have one observer:

```c++
WM_PUBLIC_EXPORT void SetActivationChangeObserver(
    aura::Window* window,
    ActivationChangeObserver* observer);
WM_PUBLIC_EXPORT ActivationChangeObserver* GetActivationChangeObserver(
    aura::Window* window);
```

Are we certain that there's not some another observer being overwriting here?

Reply:
Our `AtomDesktopNativeWidgetAura` is a subclass of the original `views::DesktopNativeWidgetAura`, which in the contructor [calls](https://cs.chromium.org/chromium/src/ui/views/widget/desktop_aura/desktop_native_widget_aura.cc?l=251) `SetActivationChangeObserver` with the `content_window_` and `this`. In our case, we call it with the same thing, since `GetNativeWindow` [returns `content_window_`](https://cs.chromium.org/chromium/src/ui/views/widget/desktop_aura/desktop_native_widget_aura.cc?l=400), and `this` still behaves the same as before (except for the fact that the subclass method gets called now in case of `OnWindowActivated`, where we call the original handler at the beginning to stay compatible).

According to the call hierarchy part of the code search site, `SetActivationChangeObserver` is only called from `NativeWidgetAura` (which we don't disturb) and from `DesktopNativeWidgetAura`, which we correctly handle. I don't see any way that we could overwrite another observer here, because we only effectively overwrite with the same class instance.